src/transfers-history/services/FundsDepositedEventsQueryService.ts

```diff
@@ -0,0 +1,173 @@
+import { SpokePool__factory } from "@across-protocol/contracts-v2";
+import { FundsDepositedEvent } from "@across-protocol/contracts-v2/dist/typechain/SpokePool";
+import { BigNumber, providers } from "ethers";
+import { ChainId, SPOKE_POOLS } from "../../constants";
+import { Logger } from "../adapters/logger";
+import { ISpokePoolContractEventsQuerier, SpokePoolEventsQuerier } from "../adapters/web3";
+import { clientConfig } from "../config";
+import { State, Transfer } from "../model/state";
+
+export interface IDepositEventsQueryServiceFactory {
+  getService(
+    state: State,
+    chainId: ChainId,
+    refChainId: ChainId,
+    blockRangeSize?: number,
+    depositorAddr?: string
+  ): IDepositEventsQueryService;
+}
+
+export class DepositEventsQueryServiceFactory implements IDepositEventsQueryServiceFactory {
+  getService(state: State, chainId: ChainId, refChainId: ChainId, blockRangeSize?: number, depositorAddr?: string) {
+    const provider = new providers.JsonRpcProvider(clientConfig.web3Providers[chainId]);
+    const spokePool = SpokePool__factory.connect(SPOKE_POOLS[chainId], provider);
+    const eventsQuerier = new SpokePoolEventsQuerier(spokePool, blockRangeSize);
+    return new DepositEventsQueryService(
+      chainId,
+      state,
+      provider,
+      eventsQuerier,
+      refChainId,
+      blockRangeSize,
+      depositorAddr
+    );
+  }
+}
+
+type GetEventsResult = {
+  fromBlock: number;
+  toBlock: number;
+  numberOfEvents: number;
+};
+
+/**
+ * Interface implemented by the classes that are used to:
+ * 1. query FundsDepositedEvent events
+ * 2. handle Web3 provider errors
+ * 3. save the events in the client state
+ */
+export interface IDepositEventsQueryService {
+  getEvents(toBlock?: number): Promise<GetEventsResult>;
+}
+
+export class DepositEventsQueryService implements IDepositEventsQueryService {
+  constructor(
+    private chainId: ChainId,
+    private state: State,
+    private provider: providers.Provider,
+    private eventsQuerier: ISpokePoolContractEventsQuerier,
+    private refChainId: ChainId,
+    private blockRangeSize?: number,
+    private depositorAddr?: string
+  ) {}
+
+  public async getEvents(toBlock?: number): Promise<GetEventsResult> {
+    const isRefChain = this.chainId === this.refChainId;
+
+    if (!isRefChain) {
+      let continueQueryEvents;
+      let fromBlock = 0;
+      let toBlock = 0;
+      let numberOfEvents = 0;
+
+      do {
+        continueQueryEvents = false;
+        const result = await this._getEvents(isRefChain, toBlock);
+        fromBlock = Math.min(fromBlock, result.fromBlock);
+        toBlock = Math.max(toBlock, result.toBlock);
+        numberOfEvents += result.numberOfEvents;
+
+        const latestRefTimestamp = this.state.progress[this.refChainId].latestFromTimestamp;
+        const latestTimestamp = this.state.progress[this.chainId].latestFromTimestamp;
```

Comment:
i think this kind of optimization might need to be justified more. Why not just let each chain download all its events in the background, and show whatevers available on the FE as it comes in?